minetest/minetest
# glTF model support

- Goal of the PR: Add glTF model support.
- How does the PR work? The Minetest side of things is trivial; just add glTF to the list of supported formats, and add some glTF test models & entities. (Also, I renamed an Irrlicht method which was wrongly named. That is a separate commit.)
- Does it resolve any reported issue? Closes #9673.
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)? glTF support is on the roadmap.
- If not a bug fix, why is this PR needed? What usecases does it solve? Has been explained in the issue already. Still, top points (IMO) are:
  * Lowers the bar to model creation.
  * Provides a solid basis for adding features beyond what .b3d / .x support (better materials, animations, etc.)
  * Brittle .b3d / .x toolchains need not be relied on anymore.

Depends on:

1. https://github.com/minetest/irrlicht/pull/200
2. https://github.com/JosiahWI/irrlicht/pull/14
3. https://github.com/JosiahWI/irrlicht/pull/12

(WIP, PRs 2 and 3 are pending review by Josiah.)

## To do

- [x] Move tiniergltf to a separate repo. (I'm willing to maintain this.)
- [x] Sparse accessor support (and related accessor refactoring).
- [x] Fix Irrlicht PR MinGW builds (~~either by bundling the jsoncpp amalgamation like Minetest or getting sfan to add Jsoncpp to the available system packages~~ fixed by forking jsoncpp and fetching it via `FetchContent`).
- [ ] Thorough testing and code review, resulting in fixes and refactorings.
- [ ] Warnings, like not using all nodes in a scene or embedding images (which we won't read).
- [ ] Profile to be sure that the loader has acceptable performance (highly likely IMO).
- [ ] (low priority, maybe out of scope, but also low effort) add .glb support for smaller file sizes.
- [ ] (very low-priority, likely out of scope) rewrite the base64 implementation for fun and profit.

## How to test

1. Check out & build https://github.com/appgurueu/irrlicht/tree/feat/gltf-loader-tinier-fixmes-animated.
2. Run the Irrlicht unittests: `ctest --output-on-failure` (inside the `irrlicht` dir).
3. Check out & build this PR.
4. Spawn the test entities.
5. :eye: :lips: :eye:

[Screencast from 09-01-24 00:47:45.webm](https://github.com/minetest/minetest/assets/34514239/e8c73903-d2d8-42ee-9a42-b15fbf036589)

## Trivia

![Spider Bunny](https://github.com/minetest/minetest/assets/34514239/f1d3c6b0-64fe-4dd3-865b-d3b543bd1cc4)

---

The work to make this possible was done by what I would call Minetest's "inofficial gltf working group" - Josiah, Jordan, and me. The spider model used for testing was made by archfan.
